File: src/lib/pricing.js
/**
 * Client-side ballpark pricing estimator — 50-state aware.
 *
 * Base rates use industry-standard cost ranges per square foot.
 * When a stateCode is supplied the estimate is adjusted using the
 * state's laborIndex and materialPremium from states50.js:
 *   multiplier = (laborIndex × 0.65) + (materialPremium × 0.35)
 *
 * Math basis (national baseline):
 *   Residential paving    $3.50 – $8.00 / sq ft
 *   Commercial paving     $2.50 – $6.00 / sq ft  (volume discount)
 *   Sealcoating           $0.15 – $0.35 / sq ft
 *   Crack filling         $0.40 – $1.00 / sq ft  (varies with crack density)
 *   Parking lot           $3.00 – $7.00 / sq ft
 *   Driveway              $3.50 – $7.50 / sq ft
 *   Maintenance plan      $0.20 – $0.45 / sq ft / year
 *
 * Returns { lowFmt, highFmt, disclaimer, stateNote? } or null.
 */
import { getPriceMultiplier, getStateSummary } from './states50'

const RATES = {
  paving: {
    residential: { low: 3.5, high: 8.0 },
    commercial: { low: 2.5, high: 6.0 },
    default: { low: 3.0, high: 7.0 },
  },
  sealcoating: {
    residential: { low: 0.15, high: 0.35 },
    commercial: { low: 0.12, high: 0.3 },
    default: { low: 0.12, high: 0.35 },
  },
  crackfill: {
    residential: { low: 0.4, high: 1.0 },
    commercial: { low: 0.35, high: 0.9 },
    default: { low: 0.4, high: 1.0 },
  },
  parking_lot: {
    residential: { low: 3.0, high: 7.0 },
    commercial: { low: 3.0, high: 7.0 },
    default: { low: 3.0, high: 7.0 },
  },
  driveway: {
    residential: { low: 3.5, high: 7.5 },
    commercial: { low: 3.0, high: 6.5 },
    default: { low: 3.5, high: 7.5 },
  },
  maintenance: {
    residential: { low: 0.2, high: 0.4 },
    commercial: { low: 0.18, high: 0.4 },
    default: { low: 0.2, high: 0.45 },
  },
  // General Contracting — GC fee as a cost-per-sqft proxy for managed construction
  // (actual cost is project-specific; these are ballpark managed-build figures)
  general_contracting: {
    residential: { low: 85.0, high: 250.0 },
    commercial: { low: 75.0, high: 200.0 },
    default: { low: 80.0, high: 225.0 },
  },
  // Interior Design — per sq ft of designed space (furnishings & materials excluded)
  interior_design: {
    residential: { low: 5.0, high: 18.0 },
    commercial: { low: 8.0, high: 25.0 },
    default: { low: 5.0, high: 20.0 },
  },
  // Cobblestone & Brick Paver Patios — installed sq ft
  cobblestone_pavers: {
    residential: { low: 15.0, high: 55.0 },
    commercial: { low: 18.0, high: 60.0 },
    default: { low: 15.0, high: 55.0 },
  },
  // Stone Masonry — installed sq ft of wall / patio surface
  stone_masonry: {
    residential: { low: 30.0, high: 85.0 },
    commercial: { low: 35.0, high: 100.0 },
    default: { low: 30.0, high: 90.0 },
  },
  // ── 3D Visualizer build types ─────────────────────────────────────────
  // New Construction — full build cost per sq ft (site, foundation, structure, finish)
  new_construction_residential: {
    residential: { low: 120.0, high: 320.0 },
    commercial: { low: 110.0, high: 280.0 },
    default: { low: 120.0, high: 320.0 },
  },
  // Addition / Remodel — per sq ft of new added/renovated space
  addition: {
    residential: { low: 90.0, high: 250.0 },
    commercial: { low: 100.0, high: 275.0 },
    default: { low: 90.0, high: 250.0 },
  },
  // ADU (Accessory Dwelling Unit) — detached or attached unit, turnkey
  adu: {
    residential: { low: 150.0, high: 400.0 },
    commercial: { low: 140.0, high: 350.0 },
    default: { low: 150.0, high: 400.0 },
  },
  // Commercial New Build — shell + TI per sq ft
  commercial_build: {
    residential: { low: 100.0, high: 260.0 },
    commercial: { low: 90.0, high: 240.0 },
    default: { low: 90.0, high: 260.0 },
  },
}

/** Format a dollar value: round to nearest $50 for clarity */
function roundToFifty(n) {
  return Math.round(n / 50) * 50
}

function fmt(n) {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    maximumFractionDigits: 0,
  }).format(n)
}

/**
 * @param {string}      serviceType  — e.g. 'paving', 'sealcoating'
 * @param {string}      propertyType — 'residential' | 'commercial'
 * @param {number|null} sqft         — project size in square feet
 * @param {string|null} stateCode    — optional 2-letter state abbr for regional adjustment
 * @returns {{ lowFmt, highFmt, disclaimer, stateNote?, multiplier? } | null}
 */
export function estimatePrice(serviceType, propertyType, sqft, stateCode = null) {
  const service = serviceType?.toLowerCase()
  const property = propertyType?.toLowerCase()
  const area = parseFloat(sqft)

  if (!service || !RATES[service] || !area || area <= 0) return null

  const rates = RATES[service][property] ?? RATES[service].default

  // Apply 50-state regional multiplier when a valid stateCode is provided
  const multiplier = stateCode ? getPriceMultiplier(stateCode.toUpperCase()) : 1.0
  const rawLow = rates.low * area * multiplier
  const rawHigh = rates.high * area * multiplier

  const low = roundToFifty(rawLow)
  const high = roundToFifty(rawHigh)

  const clampedLow = Math.max(low, 250)
  const clampedHigh = Math.max(high, 500)

  // Build optional state-specific note
  let stateNote = null
  if (stateCode) {
    const summary = getStateSummary(stateCode.toUpperCase())
    if (summary) stateNote = summary.pricingNote
  }

  return {
    lowFmt: fmt(clampedLow),
    highFmt: fmt(clampedHigh),
    multiplier,
    stateNote,
    disclaimer:
      'Ballpark estimate only — final price depends on site conditions, ' +
      'material costs, and access. Free on-site quote always included.',
  }
}

File: app/services/pricing.py
"""
Ballpark pricing estimator for J. Worden & Sons quote requests.

Math basis (industry-standard cost ranges):
  Residential paving          $3.50 – $8.00 / sq ft
  Commercial paving           $2.50 – $6.00 / sq ft  (volume & mobilization efficiencies)
  Sealcoating                 $0.15 – $0.35 / sq ft
  Crack filling               $0.40 – $1.00 / sq ft  (average crack density assumption)
  Parking lot                 $3.00 – $7.00 / sq ft
  Driveway                    $3.50 – $7.50 / sq ft
  Maintenance plan            $0.20 – $0.45 / sq ft / year
  General contracting         $75 – $250 / sq ft  (managed construction cost)
  Interior design             $5 – $25 / sq ft  (design fee; FF&E costs excluded)
  Cobblestone / brick pavers  $15 – $60 / sq ft  (installed)
  Stone masonry               $30 – $100 / sq ft  (installed wall/patio surface)

All estimates include a $300 mobilization floor — no job dispatches for less.
Returned figures are rounded to the nearest $50 for realistic quoting.

Verification:
  1 000 sq ft residential paving  → $3 500 – $8 000  ✓
  5 000 sq ft commercial paving   → $12 500 – $30 000 ✓
  2 000 sq ft sealcoating         → $300 – $700       ✓ (above mobilization floor)
  500 sq ft cobblestone patio     → $7 500 – $27 500  ✓
  200 sq ft stone masonry wall    → $6 000 – $17 000  ✓
"""

# ── Rate table: service → property_type → (low, high) $/sq ft ────────────────

_RATES: dict[str, dict[str, tuple[float, float]]] = {
    "paving":               {"residential": (3.50, 8.00),   "commercial": (2.50, 6.00)},
    "sealcoating":          {"residential": (0.15, 0.35),   "commercial": (0.12, 0.30)},
    "crackfill":            {"residential": (0.40, 1.00),   "commercial": (0.35, 0.90)},
    "parking_lot":          {"residential": (3.00, 7.00),   "commercial": (3.00, 7.00)},
    "driveway":             {"residential": (3.50, 7.50),   "commercial": (3.00, 6.50)},
    "maintenance":          {"residential": (0.20, 0.40),   "commercial": (0.18, 0.40)},
    # General Contracting — managed construction cost per sq ft (all-in build)
    "general_contracting":  {"residential": (85.00, 250.00), "commercial": (75.00, 200.00)},
    # Interior Design — design/coordination fee per sq ft of designed space
    "interior_design":      {"residential": (5.00, 18.00),  "commercial": (8.00, 25.00)},
    # Cobblestone & Brick Paver Patios — fully installed sq ft
    "cobblestone_pavers":   {"residential": (15.00, 55.00), "commercial": (18.00, 60.00)},
    # Stone Masonry — installed surface sq ft (wall face or patio)
    "stone_masonry":        {"residential": (30.00, 85.00), "commercial": (35.00, 100.00)},
    # ── 3D Visualizer build types ─────────────────────────────────────────────
    # New Construction — full turnkey build cost per sq ft
    "new_construction_residential": {"residential": (120.00, 320.00), "commercial": (110.00, 280.00)},
    # Addition / Remodel — per sq ft of added/renovated space
    "addition":             {"residential": (90.00, 250.00),  "commercial": (100.00, 275.00)},
    # ADU — detached or attached accessory dwelling unit, turnkey
    "adu":                  {"residential": (150.00, 400.00), "commercial": (140.00, 350.00)},
    # Commercial New Build — shell + tenant improvements per sq ft
    "commercial_build":     {"residential": (100.00, 260.00), "commercial": (90.00, 240.00)},
}

_MOBILISATION_FLOOR_LOW  = 300.0   # minimum low-end job cost
_MOBILISATION_FLOOR_HIGH = 600.0   # minimum high-end job cost
_ROUND_TO = 50                     # round to nearest $50


def _round_to_nearest(value: float, nearest: int) -> int:
    return int(round(value / nearest) * nearest)


def estimate_price(
    service_type: str,
    property_type: str,
    project_size_sqft: float,
    state_code: str | None = None,
    include_material_adjustment: bool = True,
) -> dict | None:
    """
    Return a ballpark cost range or None if the service is not recognised.

    Parameters
    ----------
    service_type               : one of the keys in _RATES
    property_type              : 'residential' | 'commercial'
    project_size_sqft          : positive float
    state_code                 : optional 2-letter state abbreviation for regional adjustment
    include_material_adjustment: when True, applies live EIA asphalt price multiplier (Feature 7)

    Returns
    -------
    {
        "low_usd":       int,
        "high_usd":      int,
        "low_fmt":       str,
        "high_fmt":      str,
        "multiplier":    float,   # state adjustment applied (1.0 = national average)
        "material_note": str,     # live material price note (Feature 7)
        "disclaimer":    str,
    }
    """
    from .state_data import get_price_multiplier  # noqa: PLC0415

    service   = (service_type  or "").lower().strip()
    property_ = (property_type or "residential").lower().strip()
    sqft      = float(project_size_sqft or 0)

    if service not in _RATES or sqft <= 0:
        return None

    rates = _RATES[service].get(property_) or _RATES[service].get("residential")
    if not rates:
        return None

    multiplier = get_price_multiplier(state_code) if state_code else 1.0

    # Feature 7: Apply live material price multiplier
    material_note = ""
    material_multiplier = 1.0
    if include_material_adjustment:
        try:
            from .material_prices import get_price_multiplier_with_materials  # noqa: PLC0415
            mat_result = get_price_multiplier_with_materials(state_code or "US", service)
            material_multiplier = mat_result.get("multiplier", 1.0)
            material_note = mat_result.get("note", "")
        except Exception:  # noqa: BLE001
            pass

    combined = multiplier * material_multiplier
    low_raw  = rates[0] * sqft * combined
    high_raw = rates[1] * sqft * combined

    low  = max(_round_to_nearest(low_raw,  _ROUND_TO), int(_MOBILISATION_FLOOR_LOW))
    high = max(_round_to_nearest(high_raw, _ROUND_TO), int(_MOBILISATION_FLOOR_HIGH))

    def fmt(n: int) -> str:
        return f"${n:,}"

    return {
        "low_usd":        low,
        "high_usd":       high,
        "low_fmt":        fmt(low),
        "high_fmt":       fmt(high),
        "multiplier":     round(multiplier, 3),
        "material_note":  material_note,
        "disclaimer": (
            "Ballpark estimate only — final price depends on site conditions, "
            "material costs, and access. A free on-site quote is always included."
        ),
    }
